### ISSUE TITLE
CSP as a http leak protection

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-browser#2691]
follow up from #2648 - use CSP

**Comment by seisvelas** [comment]
If my understanding is correct, the current functionality is to look for a style attribute, then check if the style attribute uses either `url() or @import`, and if so, remove it. I do not believe it is possible to recreate that functionality entirely through the use of a CSP. What a CSP **could** do is prohibit *any* style attributes from being added by users. 

For a good reference on what a CSP can do pertaining to styles, I recommend checking out MDN's excellent article on the `style-src` CSP directive: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src.

If we want to let users continue to define custom styles (and merely ban the use of `url()` and `@import`), our current approach is actually not bad. On the other hand, if we want to prohibit user-defined style attributes entirely, then we could enshrine that into our CSP.

**Comment by tomholub** [comment]
Is there any CSP we can use wrt request destination? If we limit where the app is allowed to make requests to, then even if there was a http leak, the browser would block it.

That would require us to indeed change WKD to go through backend.

**Comment by tomholub** [comment]
Forbidding CSS is difficult - if you mean inline css - we do use some dynamic height and similar.

**Comment by seisvelas** [comment]
> Is there any CSP we can use wrt request destination?

After considerable investigation, this does not appear to be possible. A CSP is about policing the *source* of a resource (style, script, image, etc), it does not have a mechanism for considering destinations of requests made within a resource. 

That is not to say there aren't other good reasons for us to implement a CSP. For one example among many, the `upgrade-insecure-requests` directive to prevent loading of any non-HTTPS resources is nice to have enabled.

**Comment by tomholub** [comment]
> That is not to say there aren't other good reasons for us to implement a CSP. For one example among many, the upgrade-insecure-requests directive to prevent loading of any non-HTTPS resources is nice to have enabled.

Let's start with enabling this. Also please review that inline javascript is disabled, or that "tagged" javascript is only possible from same origin.

After that, we can investigate other ways in which CSP can be used to our advantage. Unfortunately, CSS won't be so easy.

**Comment by seisvelas** [comment]
> Let's start with enabling this. Also please review that inline javascript is disabled, or that "tagged" javascript is only possible from same origin.

I tried to bypass the current CSP and it successfully blocked all my attempts to inject inline JS (I was trying this on the Encrypted Inbox page). If any workarounds occur I'll bring it up here.

**Comment by tomholub** [comment]
@seisvelas I have just merged #2774 which does the upgrade-insecure-requests.

Anything else we should improve in the CSP?

**Comment by seisvelas** [comment]
>Anything else we should improve in the CSP?

Yes, a bunch. I'll go through the directives listed on https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy one by one and evaluate each one to see if it's useful for us

**Comment by seisvelas** [comment]
Okay, I've gone through all of the documented CSP directives and divided them into those we may wish to use, and those we likely won't want. I include a description of the directive and how it applies to us. First, the useful ones:

## Possibly useful CSP directives for flowcrypt-browser

* connect-src: Restricts the URLs which can be loaded using script interfaces (like Fetch and XMLHttpRequest).

Only helps, why not? If someone did manage to get some hostile JS running, we could mitigate their impact with this directive.

* default-src: Serves as a fallback for the other fetch directives.

This would be nice in case a new directive is released, it can fallback to this, or in case we neglect to add one of the fetch directives for whatever reason. Better safe than sorry.

* font-src: Specifies valid sources for fonts loaded using @font-face.

Unless we want to allow arbitrary fonts loaded using @font-face (which may be the case for PGP/MIME purposes or something, if so let me know and I'll exclude this one), then this only helps.

* frame-src: Specifies valid sources for nested browsing contexts loading using elements such as <frame> and <iframe>.

We already have frame-ancestors which specifies who we can be framed in, this specifies what hosts we will frame in us. If we don't need to allow arbitrary subframes then I recommend adding this.

* img-src: Specifies valid sources of images and favicons.

I'm very uncertain whether this is an option for us, but if so we could ensure images are being loaded from the expected sources. 

* manifest-src: Specifies valid sources of application manifest files.

Pretty sure we don't need to serve arbitrary manifest files, so this should be something we can set to none (correct me if I'm wrong, though!)

* media-src: Specifies valid sources for loading media using the <audio> , <video> and <track> elements.

Seems like another example of a directive that should safely be set to none.

* object-src: Specifies valid sources for the <object>, <embed>, and <applet> elements.

The elements controlled by this directive are considered deprecated anyway, this should be set to none

* style-src: Specifies valid sources for stylesheets.

Should be set to self and inline styles moved to their own files, but maybe we need this open for PGP/MIME or something like that. Let me know your thoughts.

* worker-src: Specifies valid sources for Worker, SharedWorker, or ServiceWorker scripts.

Do we use workers? If not, we can set this to none, otherwise we can figure out what hostnames we'd need to allow. 

* base-uri: Restricts the URLs which can be used in a document's <base> element.

Do we use the <base> element in any capacity? If not, we could set this to none. 

* form-action: Restricts the URLs which can be used as the target of a form submissions from a given context.

I don't think we need any form actions that submit to a different host, so we should codify that with this directive.

* report-to: Used together with the Report-To HTTP response header field instructs the user agent to store reporting endpoints for an origin.

Enables the browser to report CSP violations as they happen. Possibly useful for finding/diagnosing bugs. Setting this up will take more than just adding a CSP directive, so if we do want this, we'd probably do it after the rest are up and working as a separate issue.

* report-uri: Instructs the user agent to report attempts to violate the Content Security Policy. These violation reports consist of JSON documents sent via an HTTP POST request to the specified URI.

This is meant to be deprecated by the previous directive, report-to. However, report-to has not yet seen implementation across major browsers, so in the meantime it's common to include this older directive as well, which serves a similar purpose.

## Non-applicable directives

* child-src: Defines the valid sources for web workers and nested browsing contexts loaded using elements such as <frame> and <iframe>.

This is for if you want to apply a uniform policy for frames and web workers. In our case, we'd probably be better off doing these separately, via the frame-src and worker-src directives. 

* prefetch-src: Specifies valid sources to be prefetched or prerendered.

Do we use prefetch or prerender? In either case, we would want to encode that via this CSP directive. Unfortunately, this is an experimental API, and is not yet supported on any major browsers. Therefore, we should hold off on this directive for now.

* plugin-types: Restricts the set of plugins that can be embedded into a document by limiting the types of resources which can be loaded.

We don't use the elements governed by this directive, so we'd be better off disabling them entirely via the object-src directive.

* sandbox: Enables a sandbox for the requested resource similar to the <iframe> sandbox attribute.

We aren't a sandbox. If we enable this, we'd have to disarm almost all of the protections it offers anyway, and the few remaining benefits can be achieved from other directives anyway.

* navigate-to: Restricts the URLs to which a document can initiate navigation by any means, including <form> (if form-action is not specified), <a>, window.location, window.open, etc.

Emails contain links, so I don't think we'd be able to make this one work for us.

* block-all-mixed-content: Prevents loading any assets using HTTP when the page is loaded using HTTPS.

When upgrade-insecure-requests is set, this directive is ignored. Since we do use upgrade-insecure-requests already, there's no point in adding this directive.

---

Let me know your thoughts on these recommendations (whether their are any you disagree should be added or excluded) and I'll figure out the best specific values for the directives we choose and make a PR adding them all.

**Comment by tomholub** [comment]
That's quite a loadful. Let's start with the ones that we can do for sure:
 - font-src (none)
 - manifest-src (none/self)
 - media-src (none)
 - style-src (just self)
 - worker-src (self)
 - form-action (none)
